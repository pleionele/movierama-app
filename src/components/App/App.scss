@import '../../styles/color.scss';

.app {
  background-color: $dark-red;
  padding: 2em;
  text-align: center;
  box-sizing: border-box;
  &__text {
    font-size: 30px;
  }
}

.layout {
  box-sizing: border-box;
  display: flex;
  flex: initial;
  flex-direction: row;
  flex-wrap: wrap;
  margin-left: 4px;
  justify-content: center;
  align-items: center;
  width: 100%;
}
