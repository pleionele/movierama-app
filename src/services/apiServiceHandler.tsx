import { getMovieSearchResults } from '../api/get-movie-search-results';
import { getPlayNowMovies } from '../api/get-playnow-movies';

export const apiServiceHandler = (searchValue = '', page = 1) => {
  if (searchValue && searchValue.length > 0) {
    return getMovieSearchResults(searchValue, page);
  }
  return getPlayNowMovies(page);
};
