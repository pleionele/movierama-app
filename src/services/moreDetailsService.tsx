import { getMovieReviews } from '../api/get-movie-reviews';
import { getMovieSimilar } from '../api/get-movie-similar';
import { getMovieVideo } from '../api/get-movie-video';

export const moreDetailsService = (id: number) => {
  const idToString = id.toString();
  const reviews: any = getMovieReviews(idToString);
  const similarMovies: any = getMovieSimilar(idToString);
  const video: any = getMovieVideo(idToString);

  return Promise.all([reviews, similarMovies, video]).then(results => {
    return Object.assign(results[0], results[1], results[2]);
  });
};
