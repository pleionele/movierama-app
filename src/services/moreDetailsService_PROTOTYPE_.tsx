import { getMovieReviews } from '../api/get-movie-reviews';
import { getMovieSimilar } from '../api/get-movie-similar';
import { getMovieVideo } from '../api/get-movie-video';
import { MoreDetails } from '../domain/movie';

export const moreDetailsService = (id: number): Promise<MoreDetails> => {
  const idToString = id.toString();
  const reviews: any = getMovieReviews(idToString);
  const similarMovies: any = getMovieSimilar(idToString);
  const video: any = getMovieVideo(idToString);

  return Promise.all([reviews, similarMovies, video]).then(results => {
    return Object.defineProperties(
      {},
      {
        reviews: { value: reviewsAdapter(results[0]) },
        similarMovies: { value: similarMoviesAdapter(results[1]) },
        video: { value: trailerAdapter(results[2]) },
      }
    );
  });
};

const similarMoviesAdapter = (movies: any) => {
  try {
    if (movies.results.length > 0) {
      // We show up up to 6
      return movies.results.slice(0, 6);
    }
    return [];
  } catch (e) {
    return [];
  }
};

const trailerAdapter = (movies: any) => {
  try {
    if (movies.results.length > 0) {
      const trailers = movies.results.map((item: any, i: number) => {
        if (item.type === 'Trailer') {
          return item.key;
        }
      });
      return trailers.length > 0 ? trailers[0] : undefined;
    }
  } catch (e) {
    return undefined;
  }
};
const reviewsAdapter = (reviews: any) => {
  try {
    if (reviews.results.length > 0) {
      // We show up 2 reviews
      return reviews.results.slice(0, 2);
    }
    return [];
  } catch (e) {
    return [];
  }
};
