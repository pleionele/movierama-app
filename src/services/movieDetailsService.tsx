import { getMovieReviews } from '../api/get-movie-reviews';
import { getMovieSimilar } from '../api/get-movie-similar';
import { getMovieVideo } from '../api/get-movie-video';

export const movieDetailsService = (id: string) => {
  const reviews: any = getMovieReviews(id);
  const similarMovies: any = getMovieSimilar(id);
  const video: any = getMovieVideo(id);

  return Promise.all([reviews, similarMovies, video]).then(results => {
    return Object.assign(results[0], results[1], results[2]);
  });
};
