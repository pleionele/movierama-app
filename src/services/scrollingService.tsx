import { getMovieSearchResults } from '../api/get-movie-search-results';
import { getPlayNowMovies } from '../api/get-playnow-movies';

export const scrollingService = (searchValue = '', page: number) => {
  if (searchValue && searchValue.length > 0) {
    return getMovieSearchResults(searchValue, page);
  }
  return getPlayNowMovies(page);
};
