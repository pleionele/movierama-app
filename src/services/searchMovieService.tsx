import { getMovieSearchResults } from '../api/get-movie-search-results';

// this service will receive the search
// and sanitize the data

export const searchMovieService = (searchValue: string) => {
  const searchResults: any = getMovieSearchResults(searchValue);
  // data sanitazation will happen here

  return searchResults;
};
