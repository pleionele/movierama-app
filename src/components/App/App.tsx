import * as React from 'react';
import './App.scss';
import { playNowMoviesService } from '../../services/playNowMovieService';
import { MovieList } from '../MovieList/MovieList';
import { SearchBox } from '../Searchbox/SearchBox';
import debounce from 'lodash.debounce';
import { apiServiceHandler } from '../../services/apiServiceHandler';

export default class App extends React.Component<any, any> {
  constructor(props: any) {
    super(props);
    this.state = {
      movieResults: [],
      movieGenres: [],
      page: 1,
      loading: false,
      searchInput: '',
      // TODO add total pages
    };
  }

  async componentDidMount() {
    window.addEventListener('scroll', this.handleScroll);

    const results = await playNowMoviesService();
    // tslint:disable-next-line: no-unused-expression
    results &&
      results.results &&
      this.setState({
        movieResults: results.results,
        movieGenres: results.genres,
      });
  }

  handleScroll = () => {
    if (
      window.innerHeight + document.documentElement.scrollTop >=
      document.documentElement.offsetHeight
    ) {
      return this.fetchMoreMovies();
    }
  };

  fetchMoreMovies = async () => {
    this.setState({ loading: true });

    const { movieResults, page, searchInput } = this.state;
    const newMovies = await apiServiceHandler(searchInput, page + 1);
    if (newMovies) {
      const allmovies = movieResults.concat(newMovies.results);
      this.setState({
        movieResults: allmovies,
        loading: false,
        page: page + 1,
      });
    }
  };

  searchHandler = debounce(async (searchInput: string) => {
    this.setState({ searchInput });

    const searchResults = await apiServiceHandler(searchInput);
    if (searchResults) {
      this.setState({
        page: searchResults.page,
        movieResults: searchResults.results,
      });
    }
  }, 500);

  public render() {
    const { movieResults, movieGenres, loading, searchInput } = this.state;
    return (
      <div className="page">
        <header className="section header__container">
          <h1 className="header">MovieRama</h1>
          <SearchBox searchHandler={this.searchHandler} />
        </header>
        <div className="movie__container" data-testid="appComponent">
          <div className="movie__grid">
            {movieResults && (
              <MovieList apiResults={movieResults} movieGenres={movieGenres} />
            )}
          </div>
          {loading && <div>Please wait...</div>}
        </div>
      </div>
    );
  }
}
