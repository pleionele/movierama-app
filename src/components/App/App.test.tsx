import App from './App';
import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
// @ts-ignore
import * as payloadPayNowMovie from '../../api/stubs/payload-pay-now-movie.json';
import { playNowMoviesService } from '../../services/playNowMovieService';
import { searchMovieService } from '../../services/searchMovieService';
import debounce from 'lodash.debounce';
import { act } from 'react-dom/test-utils';

jest.mock('../../services/playNowMovieService', () => {
  return {
    playNowMoviesService: jest.fn(),
  };
});

jest.mock('../../services/searchMovieService', () => {
  return {
    searchMovieService: jest.fn(),
  };
});

describe('<App />', () => {
  beforeEach(() => {
    (debounce as jest.Mock) = jest.fn(searchTerm => {
      searchMovieService(searchTerm);
    });
    (playNowMoviesService as jest.Mock).mockReturnValue(payloadPayNowMovie);
  });

  afterEach(() => {
    (debounce as jest.Mock).mockClear();
  });

  it('should display', () => {
    // @ts-ignore
    playNowMoviesService.mockReturnValue(payloadPayNowMovie);
    const { getByTestId } = render(<App />);
    expect(getByTestId('appComponent')).toBeTruthy();
  });

  it('should have a search box', () => {
    // @ts-ignore
    playNowMoviesService.mockReturnValue(payloadPayNowMovie);
    const { getByTestId } = render(<App />);
    expect(getByTestId('appSearchBox')).toBeTruthy();
  });

  it('should match the snapshot', () => {
    act(async () => {
      const { asFragment } = render(<App />);
      await Promise.resolve();
      expect(asFragment()).toMatchSnapshot();
    });
  });
  // it('should show results from the search api when the user tries to search', () => {
  //   const { getByTestId } = render(<App name="world" />);
  //   // expect(searchMovieService).toHaveBeenCalledTimes(0);

  //   const searchInput = getByTestId('appSearchBox');
  //   fireEvent.change(searchInput, { target: { value: 'test' } });
  //   expect(debounce).toHaveBeenCalled();
  // });
});
