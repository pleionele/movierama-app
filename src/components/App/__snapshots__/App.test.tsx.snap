// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="page"
  >
    <header
      class="section header__container"
    >
      <h1
        class="header"
      >
        MovieRama
      </h1>
      <input
        aria-label="searchInput"
        class="search"
        data-testid="appSearchBox"
        placeholder="Search movies"
        type="search"
      />
    </header>
    <div
      class="movie__container"
      data-testid="appComponent"
    >
      <div
        class="movie__grid"
      />
    </div>
  </div>
</DocumentFragment>
`;
